kuma/attachments/models.py

```diff
@@ -2,16 +2,39 @@
 from datetime import datetime
 
 from django.conf import settings
+from django.core.files.storage import default_storage
 from django.db import models
 from django.db.utils import IntegrityError
 from django.utils.encoding import python_2_unicode_compatible
 from django.utils.functional import cached_property
 from django.utils.translation import ugettext_lazy as _
 from django_mysql.models import Model as MySQLModel
+from storages.backends.s3boto3 import S3Boto3Storage
 
 from .utils import attachment_upload_to, full_attachment_url
 
 
+if settings.ATTACHMENTS_USE_S3:
+    storage = S3Boto3Storage(
```

Comment:
> We don't want that. Attachments are a special case; they should specify their own storage system.

Mind you, `models.FileField` is only used by `kuma/attachments/models.py`. 

Suppose that we someday add something elsewhere we have files coming in. Wouldn't we want that on S3 too? We definitely don't want regular fs disk in this era of clouds. Yes, perhaps it *could* be a thing to have some stuff uploaded to GCS and some to S3 but it's unlikely. 

I'm too lazy to read up but surely there's a way with `django-storages` to control the destinations for different ORM things.

All in all, having migrations that depend on overrideable config seems a bigger negative than to have a broad default storage solution for all models. What do you think?

Reply:
Looks like one way to do this is to create a subclass of S3BotoStorage and override the bucket name. Something like: 

```
class MyS3Storage(S3BotoStorage):
    def __init__(self, *args, **kwargs):
        kwargs['bucket'] = getattr(settings, 'MY_AWS_STORAGE_BUCKET_NAME')
        super(MyS3Storage, self).__init__(*args, **kwargs)
```

Then you can set `DEFAULT_FILE_STORAGE` to that subclass.

Reply:
> Attachments are a special case; they should specify their own storage system.

Implementing how I mentioned I suppose you could have a subclass of `AttachmentS3Storage` that overrides what is necessary to specify it's own storage without having to grab the bucket name from the ENV?

Reply:
> Mind you, `models.FileField` is only used by `kuma/attachments/models.py`.

We use Boto and access S3 elsewhere in Kuma, unrelated to attachments, and we do so using global environment variables. Attachments needs different settings from those.

> Suppose that we someday add something elsewhere we have files coming in. Wouldn't we want that on S3 too?

We almost certainly don't want it to be in the exact same bucket, with the exact same object properties, and the exact same CDN configuration as for attachments. So we're going to have to special case something somewhere.

> I'm too lazy to read up but surely there's a way with `django-storages` to control the destinations for different ORM things.

If you're going to push this point, please do read the docs or present a working alternative. Django's defined extension point for file operations is passing differently configured Storage instances to specific models, or changing the global default.

Reply:
@joedarc's idea of creating a S3Boto3Storage subclass and building the configuration in its constructor sounds promising. I'll give that a shot and see if it improves the migration.

Reply:
What I meant was that there is no other use of `models.FileField` in any of the other apps. The only django app that have that uses file storage is [the `attachment` one](https://github.com/mozilla/kuma/tree/af446fe757a1fe87f80e452179ab734b33d0fb06/kuma/attachments).

Also, `ripgrep` only finds:

```
▶ rg FileField
kuma/attachments/models.py
106:    file = models.FileField(upload_to=attachment_upload_to, max_length=500)
216:    file = models.FileField(

kuma/attachments/migrations/0001_squashed_0008_attachment_on_delete.py
34:                ('file', models.FileField(max_length=500, upload_to=kuma.attachments.utils.attachment_upload_to)),
80:                ('file', models.FileField(help_text='The attachment file that was trashed', max_length=500, upload_to=kuma.attachments.utils.attachment_upload_to)),
```

Actually, it appears that `AWS_STORAGE_BUCKET_NAME` rules and there's no easy way to have a different bucket for a different `FileField`. But perhaps it's moot because I'm guessing `django-storages` gives the keys sane name. 

So, to avoid the whole migration problem, if we just set
```python
DEFAULT_FILE_STORAGE = 'storages.backends.s3boto3.S3Boto3Storage'
```
in `settings/common.py` 
the only thing that needs to be done is to set the various [AWS S3 related settings](https://django-storages.readthedocs.io/en/latest/backends/amazon-S3.html). 
And if any of that clashes with how we use AWS and S3 for other stuff, you'd change it to something like

```python
DEFAULT_FILE_STORAGE = 'kuma.core.storages.KumaFileStorage'
```
and `kuma/core.storages.py` something like this:
```python
from storages.backends.s3boto3 import S3Boto3Storage


class KumaFileStorage(S3Boto3Storage):
    # something like @joedarc suggested
```

That should avoid the migrations problem. It should also give you full control over the environment variables. 

Also, perhaps by doing that we'd gain some other capabilities to better control the uploads and stuff depending on prod or local dev in docker-compose.

Reply:
@peterbe How do you feel about https://github.com/mozilla/kuma/pull/5998/commits/3668f2c00d807999c14c309ef7c097da93b5a6b2? (@joedarc's suggestion was great!)

And do you still have a strong preference for `DEFAULT_FILE_STORAGE`? It would avoid a migration altogether, but at the risk of implicitly throwing future FileField/ImageField/etc's into the same bucket as attachments.

I prefer to keep it totally explicit to avoid that risk, but I could be convinced.